[SPARK-27945][SQL] Minimal changes to support columnar processing

revans2 opened the pull request:
## What changes were proposed in this pull request?

This is the first part of [SPARK-27396](https://issues.apache.org/jira/browse/SPARK-27396).  This is the minimum set of changes necessary to support a pluggable back end for columnar processing.  Follow on JIRAs would cover removing some of the duplication between functionality in this patch and functionality currently covered by things like ColumnarBatchScan.

## How was this patch tested?

I added in a new unit test to cover new code not really covered in other places.

I also did manual testing by implementing two plugins/extensions that take advantage of the new APIs to allow for columnar processing for some simple queries.  One version runs on the [CPU](https://gist.github.com/revans2/c3cad77075c4fa5d9d271308ee2f1b1d).  The other version run on a GPU, but because it has unreleased dependencies I will not include a link to it yet.

The CPU version I would expect to add in as an example with other documentation in a follow on JIRA

This is contributed on behalf of NVIDIA Corporation.

revans2 committed:
SPARK-27945: Minimal changes to support columnar processing

AmplabJenkins commented:
Can one of the admins verify this patch?

tgravescs commented:
ok to test

AmplabJenkins commented:
Can one of the admins verify this patch?

AmplabJenkins commented:
Can one of the admins verify this patch?

SparkQA commented:
**[Test build #106161 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/106161/testReport)** for PR 24795 at commit [`40c8199`](https://github.com/apache/spark/commit/40c8199906ac3629bb9e611d03735169e35179b9).

SparkQA commented:
**[Test build #106161 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/106161/testReport)** for PR 24795 at commit [`40c8199`](https://github.com/apache/spark/commit/40c8199906ac3629bb9e611d03735169e35179b9).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/106161/
Test FAILed.

revans2 commented:
Just as a high level overview of what the changes are.

The classes in vectorized, including ColumnVector, and ColumnarBatch were moved to catalyst from core so that expressions could get access to them.  To be able to comply with conventions about releasing a ColumnVector's resources we also made ColumnVector reference counted, but restrict the usage to really just BoundReference. 

Expression was updated to include optional columnar evaluation that is off by default, as was SparkPlan.  A few expressions were updated to also do columnar evaluation, but these are things like literals that there really is only one way to do it.

Columnar.scala is where the core of the changes are.  It adds in a new rule that runs on the physical plan during the execution stage. It will look at the plan and find sections that should be run columnar and will insert transitions to/from columnar formatted data to allow for the processing to happen. It can also take a user supplied set of rules to also modify the plan so that an extension can add in columnar versions of operations and expressions themselves.

I also updated the Extensions API to allow for the rules to be injected.

revans2 committed:
Fixed test style issue

SparkQA commented:
**[Test build #106162 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/106162/testReport)** for PR 24795 at commit [`484fb41`](https://github.com/apache/spark/commit/484fb4194f3a4d1c83a0d9b041521cff1dd01013).

SparkQA commented:
**[Test build #106162 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/106162/testReport)** for PR 24795 at commit [`484fb41`](https://github.com/apache/spark/commit/484fb4194f3a4d1c83a0d9b041521cff1dd01013).
 * This patch **fails MiMa tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/106162/
Test FAILed.

gatorsmile commented:
cc @mengxr @rxin @rednaxelafx

dongjoon-hyun commented:
cc @dbtsai

dongjoon-hyun commented on NOTICE-binary:
```diff
@@ -73,6 +73,10 @@ Copyright 2005-2015 The Apache Software Foundation
+This product includes software developed at
+NVIDIA (https://www.nvidia.com)
+* Copyright 2019 NVIDIA CORPORATION
+
```
Ur, what does bring this `NOTICE-binary`?

revans2 replied:
The lawyers at NVIDIA require that we place a copyright notice on contributions that are nontrivial. This is the first "significant" contribution to Apache Spark from NVIDIA so I am adding a copyright notice following the guidelines from ASF at https://www.apache.org/legal/src-headers.html#notice I followed the example of the lines right above it and it should be generic enough to cover any future contribution, so this is a one-time thing.

rxin replied:
Is there a precedent on this before? My understanding is that the software is developed at the Apache Software Foundation, not the contributor's company.

revans2 replied:
I totally agree that this is not common.  I am not a lawyer and personally I agree with your interpretation, but the NVIDIA lawyers want the copyright notice in some form.  I can try and escalate it and see what has happened with NVIDIA for other ASF projects, but in the mean time if you could look at the rest of the patch that would be great.

rxin replied:
Will do. Thanks.

dongjoon-hyun replied:
Ping, @srowen . How do you think about this?

IIUC, since this PR is submitted directly to the ASF Apache Spark project, this cannot be a `NOTICE` for 3rd party work.
- https://www.apache.org/legal/src-headers.html#notice
- https://www.apache.org/legal/src-headers.html#3party
`The term "third-party work" refers to a work not submitted directly to the ASF by the copyright owner or owner's agent.`

Anyway, I understand that PMC can make a decision case-by-case.

srowen replied:
Yes, you would not be able to put this copyright message in an ASF project for code that was contributed to the project (vs code we include from third parties). It's not that it isn't true, it's that it's more specifically licensed to the ASF. We can ask for a CCLA if necessary but all the contributions would need to be under the AL2 license.

revans2 replied:
Okay I have spoken with some of the legal team here and I have a better understanding of the issue, and an alternative solution. They are not really concerned about protecting NVIDIA by this, but instead want to make sure that there is clear traceability that this contribution is on behalf of NVIDIA and covered by their corporate CLA. From what I was told this is primarily to give our end users enough information that they can apply the Apache License properly especially in relation to the patent grant clause.

I am not a lawyer, this is just my understanding of what was said.  So I will be removing the change to this file, and instead I'll just make it clear in this PR, and any other PR I do going forward that is being contributed by on on behalf of my employer (NVIDIA).

dongjoon-hyun replied:
Thank you for removing this, @revans2 .

rxin commented:
@revans2 why do we need a ColumnarRule? It seems like you are bifurcating everything.

rxin commented on sql/core/src/main/scala/org/apache/spark/sql/execution/Columnar.scala:
```diff
@@ -0,0 +1,629 @@
+ * inserting stages to build larger batches for more efficient processing, or stages that
+ * transition the data to/from an accelerator's memory.
+ */
+@Experimental
```
all apis in execution are non-public so there's no experimental api here.

revans2 replied:
So how do we want to handle this situation where we are exposing some internal implementation details but with no real guarantees of stability?  Not marking them as anything I felt gave the wrong impression that there was some guarantee.

rxin replied:
They are all private. The package.scala file documents that. Also execution and catalyst packages are ignored in the public API docs.

Basically it's use at your own risk type of internal APIs.

revans2 commented:
@rxin we don't need a ColumnarRule.  It just made the code simpler. I see your point and I am happy to change it if you like.  I can do separate registration for pre/post rules, or I can just have them be internal to ColumnarRule.
```
class ColumnarRule {
  def pre: Rule[SparkPlan] = plan => plan
  def post: Rule[SparkPlan] = plan => plan
}
```

revans2 committed:
Addressed some initial review comments

SparkQA commented:
**[Test build #106205 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/106205/testReport)** for PR 24795 at commit [`52e31ab`](https://github.com/apache/spark/commit/52e31ab60cae326f42796923875a1c001a53161a).

SparkQA commented:
**[Test build #106205 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/106205/testReport)** for PR 24795 at commit [`52e31ab`](https://github.com/apache/spark/commit/52e31ab60cae326f42796923875a1c001a53161a).
 * This patch **fails MiMa tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/106205/
Test FAILed.

revans2 committed:
Merge branch 'master' into columnar-basic

SparkQA commented:
**[Test build #106242 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/106242/testReport)** for PR 24795 at commit [`2a3287b`](https://github.com/apache/spark/commit/2a3287b4b9e241bbbdae74f4ae6c9cb390550de1).

SparkQA commented:
**[Test build #106242 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/106242/testReport)** for PR 24795 at commit [`2a3287b`](https://github.com/apache/spark/commit/2a3287b4b9e241bbbdae74f4ae6c9cb390550de1).
 * This patch **fails MiMa tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/106242/
Test FAILed.

kiszk commented on sql/catalyst/src/main/java/org/apache/spark/sql/vectorized/ColumnVector.java:
```diff
@@ -50,19 +50,44 @@
+   * Actually cleans up memory for this column vector. The column vector is really not usable after
+   * this.
+   */
+  protected abstract void doClose();
```
Can we use self-declarative name? This is because `columnVector` is public API for developers who want to support their storage.

kiszk commented on sql/core/src/main/scala/org/apache/spark/sql/execution/Columnar.scala:
```diff
@@ -0,0 +1,609 @@
+ * inserting stages to build larger batches for more efficient processing, or stages that
+ * transition the data to/from an accelerator's memory.
+ */
+class ColumnarRule {
```
Are `pre` and `post` appropriate naming, in particular, `pre`?  IIUC correctly, `pre` is a mandatory function for columnar processing and `post` is an optional function to optimize columnars or to clean up resources. To me, a pair of `pre` and `post` looks dual. (Of course, I am not a native English speaker).

In addition, can you create a test case using `ColunarRule` (for example, a simplified version of [this](https://gist.github.com/revans2/c3cad77075c4fa5d9d271308ee2f1b1d))? It would help reviewers and can ensure the behavior of this API.

revans2 replied:
For me `pre` and `post` made since in relation to the comments, but perhaps not standalone. I'll see if I can make them more descriptive.  I'll also add in a test.

revans2 committed:
removed unneeded change

SparkQA commented:
**[Test build #106255 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/106255/testReport)** for PR 24795 at commit [`31d7248`](https://github.com/apache/spark/commit/31d72489d1aed65e5616a779fd77be451a9817d9).

SparkQA commented:
**[Test build #106255 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/106255/testReport)** for PR 24795 at commit [`31d7248`](https://github.com/apache/spark/commit/31d72489d1aed65e5616a779fd77be451a9817d9).
 * This patch **fails MiMa tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/106255/
Test FAILed.

rxin commented:
@revans2 I still feel this is pretty invasive. I think you can still accomplish your goal but reduce the scope.

If I understand your use case clearly, what you want to do is to build some processing logic completely outside Spark. What you really need there is quite small (not very different from an uber UDF interface):

1. The ability to inspect and transform Catalyst plans, to create your own plans, which already exists
2. The ability for operators to read columnar input, which somewhat exists at the scan level for Parquet and ORC but is a bit hacky at the moment. It can be generalized to support arbitrary operators.
3. A conversion operator that generates ColumnarBatches.

I think that's all you need? There is no need to define column specific rules, or create a columnar expression interface. Those expressions are entirely outside Spark, and don't really need to be Spark Expression's.

revans2 commented:
@rxin

I am a confused on exactly what you are proposing.

> The ability to inspect and transform Catalyst plans, to create your own plans, which already exists
That I understand and there are some large technical burdens with using the current set of APIs, but I can get into that more if needed.

> The ability for operators to read columnar input, which somewhat exists at the scan level for Parquet and ORC but is a bit hacky at the moment. It can be generalized to support arbitrary operators.

This is the one that I don't understand what you are proposing.  The only API even remotely associated with SparkPlan where columnar formatted data can leave the plan is

https://github.com/apache/spark/blob/5cdc50684883fe1134d1261b67127c10ec452b65/sql/core/src/main/scala/org/apache/spark/sql/execution/WholeStageCodegenExec.scala#L77-L82

Which is a part of the `CodeGenSupport` trait. To look at it you would have no idea at all that it ever could actually hold `ColumnarBatch`es instead of `InternalRow`s.  This hack is really just a way so that a leaf node in the plan which pulls in `ColumnarBatchScan` can delegate to code generation the conversion between batches and rows. I'm not sure how that in particular can be generalized into something that would allow for columnar input to arbitrary operators.

As such are you suggesting then that we keep the columnar execution path for SparkPlan like in this patch, but not the Expressions?  Are you suggesting instead that we don't change SparkPlan and have it so that each operator can output an `RDD[ColumnarBatch]`, but type erase it to an `RDD[InternalRow]`?  We would still need some kind of flag or something else to indicate that this is happening in the child and a `Rule` of some kind to make sure that the transitions were inserted at the appropriate places.

rxin commented:
I'm suggesting breaking this into two changes: first change is columnar interface (this physical plan perhaps takes columns and generates columns); second change is create a converter operator.

You shouldn't need the optimizer changes or the reference counting changes, or the expression changes...

revans2 commented:
Yes, the reference counting and the expression changes could technically be removed.  The reference counting is there to be able to support columnar expressions.

In reality I believe that none of this patch is technically required to make columnar processing work as a plugin. The reason I picked this level of abstraction, and adding in new rules that modify the physical plan was to try and reduce the amount of redundant code between a plugin and Spark SQL itself to the point that it would be preferable to use the plugin API over forking Spark.  Every other implementation of columnar processing on Spark has gone the route of forking because the existing extension APIs, thought technically sufficient, require any plugin to re-implement large parts of query planning and execution.  

For example, user defined rules for mapping a logical to a physical plan run prior to the built in rules.  This means that if I add in a rule to replace a projection with a columnar enabled version the built in rule that creates the FileScanExec and does predicate push down into Parquet or Orc will not match the projection I just mapped. Predicate push down is then silently disabled.  We ran into this issue when we tried to use the existing extension APIs.  

Another example is ShuffleExchangeExec. The best performance optimization is to not do something, so if we can keep the data columnar through an exchange we have eliminated one columnar to row translation and one row to columnar translation. This is a part of the requirements in the SPIP.  Replacing that in the logical to physical plan translation is possible, although difficult.  But it also has the unintended consequence of the ReuseExchange rule not matching our alternative ShuffleExchangeExec and without full access to the physical plan I don't think we could re-implement that in the Logical to physical plan translation.

I picked a very specific place in the life cycle of the plan to inject these rules because it was as close to the final physical plan as possible so there were the fewest number of transformation afterwards.  Even then the code generation rule afterwards still required some modifications.

I think dropping the changes to the Expressions is doable, but not having a way to modify the physical plan very late in the process would require us to really rethink the direction we are taking.

revans2 committed:
Addressed more review comments and added new test

SparkQA commented:
**[Test build #106279 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/106279/testReport)** for PR 24795 at commit [`a388060`](https://github.com/apache/spark/commit/a38806018220479898ca87d5407c395a66b32e4c).

SparkQA commented:
**[Test build #106279 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/106279/testReport)** for PR 24795 at commit [`a388060`](https://github.com/apache/spark/commit/a38806018220479898ca87d5407c395a66b32e4c).
 * This patch **fails MiMa tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/106279/
Test FAILed.

revans2 committed:
Removed columnar changes to Expression

revans2 committed:
Removed reference counting from ColumnVector

SparkQA commented:
**[Test build #106357 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/106357/testReport)** for PR 24795 at commit [`ca6bb76`](https://github.com/apache/spark/commit/ca6bb7626670e2d53913eed915412db15acce2b0).

revans2 commented:
@rxin 

I removed the columnar changes to Expression and also reference counting from ColumnVector.  Please take another look.

SparkQA commented:
**[Test build #106357 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/106357/testReport)** for PR 24795 at commit [`ca6bb76`](https://github.com/apache/spark/commit/ca6bb7626670e2d53913eed915412db15acce2b0).
 * This patch **fails to generate documentation**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/106357/
Test FAILed.

revans2 committed:
Work around for docs issues

SparkQA commented:
**[Test build #106361 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/106361/testReport)** for PR 24795 at commit [`14547b3`](https://github.com/apache/spark/commit/14547b3765942dbb6936ab6fd817931d6caf3949).

SparkQA commented:
**[Test build #106361 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/106361/testReport)** for PR 24795 at commit [`14547b3`](https://github.com/apache/spark/commit/14547b3765942dbb6936ab6fd817931d6caf3949).
 * This patch **fails to generate documentation**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/106361/
Test FAILed.

revans2 committed:
Some more fixes for scaladocs in test

SparkQA commented:
**[Test build #106362 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/106362/testReport)** for PR 24795 at commit [`6c0a8ff`](https://github.com/apache/spark/commit/6c0a8ff561305f6699a9a7507dee634b911317c4).

SparkQA commented:
**[Test build #106362 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/106362/testReport)** for PR 24795 at commit [`6c0a8ff`](https://github.com/apache/spark/commit/6c0a8ff561305f6699a9a7507dee634b911317c4).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/106362/
Test PASSed.

abellina commented on sql/core/src/test/scala/org/apache/spark/sql/SparkSessionExtensionSuite.scala:
```diff
@@ -116,6 +121,23 @@ class SparkSessionExtensionSuite extends SparkFunSuite {
+      import session.sqlContext.implicits._
+      // repartitioning avoids having the add operation pushed up into the LocalTableScan
+      val data = Seq((100L), (200L), (300L)).toDF("vals").repartition(1)
+      val result = data.selectExpr("vals + 1").collect()
```
maybe lets make the rule `PreRuleReplaceAddPlusExtraOne` as I see that 100 + 1 is not 102, but you stated this within the rule :) 

re: https://github.com/apache/spark/pull/24795/files#diff-b7face523211086b889c334ae4075ad2R549

abellina reviewed (changes requested):
My comments are mostly nits, please let me know if you have questions.

revans2 committed:
Addressed review comments

abellina commented:
Thanks for addressing the review comments @revans2. 

I am +1 (non binding). I can't count myself as a catalyst expert, and so I hope we get some more folks here to take a look, but overall I think it makes sense how it fits into the existing code. I think this is a clean way to add the plugin, and the interface is clean. 

Thanks!

SparkQA commented:
**[Test build #106530 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/106530/testReport)** for PR 24795 at commit [`5815658`](https://github.com/apache/spark/commit/5815658a94d8c246725ce6f79dcdd5d092336f08).

SparkQA commented:
**[Test build #106530 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/106530/testReport)** for PR 24795 at commit [`5815658`](https://github.com/apache/spark/commit/5815658a94d8c246725ce6f79dcdd5d092336f08).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class BrokenColumnarAdd(left: ColumnarExpression, right: ColumnarExpression)`
  * `case class PreRuleReplaceAddWithBrokenVersion() extends Rule[SparkPlan] `

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/106530/
Test PASSed.

BryanCutler commented on sql/core/src/main/scala/org/apache/spark/sql/execution/Columnar.scala:
```diff
@@ -0,0 +1,594 @@
+
+  private object BooleanConverter extends TypeConverter {
+    override def append(row: SpecializedGetters, column: Int, cv: WritableColumnVector): Unit =
+      if (row.isNullAt(column)) {
```
Could `TypeConverter` take care of writing the nulls so it's not duplicated on each converter?

revans2 replied:
Possibly.  Structs need a different API to append a null, so I would have to special case those within the parent code.  I also was thinking of taking nullability into account to avoid some of the conditionals all together, but yes, I think there is the possibility of reducing the amount of code further.  I'll see what I can come up with.

BryanCutler reviewed (commented):
Looks pretty reasonable to me, I just had some minor comments. It would be nice to eventually see an example of using Arrow vectors, and share some common code with the existing Arrow functionality, but I suppose that can come later.

I noticed your example plugin defines the conf `org.apache.spark.example.columnar.enabled`, is that meant to be done per plugin or is there going to be a global conf to toggle columnar processing?

kiszk commented on sql/core/src/test/scala/org/apache/spark/sql/SparkSessionExtensionSuite.scala:
```diff
@@ -251,6 +276,368 @@ object MyExtensions {
+}
+
+case class MyPostRule() extends Rule[SparkPlan] {
+  override def apply(plan: SparkPlan): SparkPlan = plan
```
it looks identical plan. We would appreciate it if the testcase would have non-identical plan for ease of review.

revans2 replied:
Sure, I'll look at inserting in a noop of some kind to show that it was replaced.

revans2 commented on sql/core/src/main/scala/org/apache/spark/sql/execution/Columnar.scala:
```diff
@@ -0,0 +1,594 @@
+ * Provides an optimized set of APIs to extract a column from a row and append it to a
+ * [[WritableColumnVector]].
+ */
+private object RowToColumnConverter {
```
That is the eventual goal.  In this patch, I wanted the code to not replace any existing code to keep the possible impacts smaller.  

I am currently working on a separate patch to replace ColumnarBatchScan with ColumnarToRowExec.  The patch is mostly just updating tests to match the new pattern, but I also want to run a lot of benchmark code to be sure that I didn't regress anything.

After that, I was going to start working on the various APIs that use arrow data behind the scenes to process data, R, Python, etc.  That is likely to have a larger impact because the code that does those transitions is spread through a lot more files.  I also don't want to get too far ahead of myself as this patch all of them would be based off of is not even in yet.

revans2 committed:
Merge branch 'master' into columnar-basic

revans2 committed:
Addressed review comments

revans2 committed:
Addressed more review comments

revans2 committed:
Updated test to verify that rules replaced expected plan elements

revans2 committed:
Some code cleanup and better support for nullability in row to columnar conversion

revans2 commented:
@kiszk and @BryanCutler I upmerged to fix the merge conflict and I think I have addressed all of your review comments.  Please take another look.

SparkQA commented:
**[Test build #106593 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/106593/testReport)** for PR 24795 at commit [`3b30b06`](https://github.com/apache/spark/commit/3b30b06adce23ddb7105df664ae10459191272f6).

revans2 committed:
Bug fix for unsafe projection

SparkQA commented:
**[Test build #106595 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/106595/testReport)** for PR 24795 at commit [`4c01b74`](https://github.com/apache/spark/commit/4c01b74bcac397637e646f84fee67b18a8e0c316).

SparkQA commented:
**[Test build #106593 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/106593/testReport)** for PR 24795 at commit [`3b30b06`](https://github.com/apache/spark/commit/3b30b06adce23ddb7105df664ae10459191272f6).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/106593/
Test PASSed.

SparkQA commented:
**[Test build #106595 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/106595/testReport)** for PR 24795 at commit [`4c01b74`](https://github.com/apache/spark/commit/4c01b74bcac397637e646f84fee67b18a8e0c316).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/106595/
Test PASSed.

BryanCutler reviewed (commented):
Changes look good, thanks! I'm still wondering if the conf `org.apache.spark.example.columnar.enabled` or similar is meant to be per plugin or is there going to be a global conf to turn off all columnar processing?

revans2 commented:
@BryanCutler Sorry about missing that question.

Yes, the assumption is that each plugin could/would add their own configs.  We don't want to globally disable columnar support because we will eventually use it in regular operation, like with batch data being read from parquet or the transformations for pandas or R.

revans2 committed:
Addressed review comments

SparkQA commented:
**[Test build #106636 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/106636/testReport)** for PR 24795 at commit [`dc53e83`](https://github.com/apache/spark/commit/dc53e83726e2329d6a6f882ed54b7a25e7fd519f).

SparkQA commented:
**[Test build #106636 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/106636/testReport)** for PR 24795 at commit [`dc53e83`](https://github.com/apache/spark/commit/dc53e83726e2329d6a6f882ed54b7a25e7fd519f).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/106636/
Test PASSed.

kiszk commented on sql/core/src/main/scala/org/apache/spark/sql/execution/Columnar.scala:
```diff
@@ -0,0 +1,538 @@
+
+        TaskContext.get().addTaskCompletionListener[Unit] { _ =>
+          if (cb != null) {
+            cb.close
```
nit: `cb.close()`?

kiszk commented on sql/core/src/test/scala/org/apache/spark/sql/SparkSessionExtensionSuite.scala:
```diff
@@ -16,14 +16,19 @@
+import org.apache.spark.rdd.RDD
 import org.apache.spark.sql.catalyst.{FunctionIdentifier, TableIdentifier}
-import org.apache.spark.sql.catalyst.expressions.{Expression, ExpressionInfo, Literal}
+import org.apache.spark.sql.catalyst.expressions.{Add, Alias, AttributeReference, AttributeSeq, BoundReference, Expression, ExpressionInfo, ExprId, Literal, NamedExpression}
```
nit: how about ` import org.apache.spark.sql.catalyst.expressions._`. It looks too long

kiszk commented on sql/core/src/test/scala/org/apache/spark/sql/SparkSessionExtensionSuite.scala:
```diff
@@ -16,14 +16,19 @@
 import org.apache.spark.sql.catalyst.rules.Rule
-import org.apache.spark.sql.execution.{SparkPlan, SparkStrategy}
-import org.apache.spark.sql.types.{DataType, IntegerType, StructType}
+import org.apache.spark.sql.execution.{ColumnarRule, ColumnarToRowExec, ProjectExec, RowToColumnarExec, SparkPlan, SparkStrategy}
```
ditto

revans2 committed:
Addressed review comments

SparkQA commented:
**[Test build #106732 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/106732/testReport)** for PR 24795 at commit [`8443b29`](https://github.com/apache/spark/commit/8443b290f3c9a3b2272d014e183427a107e0c5e1).

revans2 commented:
@BryanCutler @kiszk @rxin 

I think I have addressed all of the review comments so far. I am happy to do more if there are more issues, but I really would like to keep this moving and try to get some of the follow on work up as well.  Please let me know what else is needed to get this merged in.

kiszk commented on sql/core/src/test/scala/org/apache/spark/sql/SparkSessionExtensionSuite.scala:
```diff
@@ -116,6 +121,34 @@ class SparkSessionExtensionSuite extends SparkFunSuite {
+      val df = data.selectExpr("vals + 1")
+      // Verify that both pre and post processing of the plan worked.
+      val found = df.queryExecution.executedPlan.collect {
+        case rep: ReplacedRowToColumnarExec => 1
```
How about the following? The summation return the unique value that corresponds to the combination.

```
case rep: ... => 1
case proj: ... => 4
case c2r: ... => 13
```

SparkQA commented:
**[Test build #106732 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/106732/testReport)** for PR 24795 at commit [`8443b29`](https://github.com/apache/spark/commit/8443b290f3c9a3b2272d014e183427a107e0c5e1).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/106732/
Test PASSed.

BryanCutler reviewed (commented):
LGTM as a first step, what will be the next task in the epic?

rxin commented on sql/core/src/main/scala/org/apache/spark/sql/execution/Columnar.scala:
```diff
@@ -0,0 +1,534 @@
+import org.apache.spark.sql.vectorized.{ColumnarBatch, ColumnVector}
+
+/**
+ * Holds a user defined rule that can be used to inject columnar implementations of various
```
I still don't understand - why do we need ColumnarRule, i.e. any rule that's column specific?

revans2 replied:
@rxin 
As I explained [before](https://github.com/apache/spark/pull/24795#issuecomment-499960263) there needs to be at least one rule that operates on the entire physical `SparkPlan` after the exchanges have been inserted into the plan.

Having two rules, one that runs prior to inserting columnar transitions and one that runs after it makes it so a plugin does not have to duplicate code for inserting columnar transitions and provides a cleaner API for anyone trying to insert columnar processing, but that is relatively minor.

If you want me to split `ColumnarRule` up and rename the individual parts to not be columnar specific I am happy to do it.  If you want me to make it a single rule that runs prior to code generation I can make that work too.  Just let me know which of the two changes you would prefer, and I will make it happen.

revans2 commented:
@BryanCutler the next step is to remove ColumnarBatchScan and replace it with the injected columnar transitions.  I have a patch for that ready to go as soon as this patch goes in.

After that, I would look at how to update the code that uses arrow vectors to communicate with Python/R/.net, etc and how best to update them to commonize as much of the code as possible.

revans2 committed:
Addressed review comments

SparkQA commented:
**[Test build #106769 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/106769/testReport)** for PR 24795 at commit [`7f1753d`](https://github.com/apache/spark/commit/7f1753db2d15186399e5ef9d7ef88b2bc7eedf67).

SparkQA commented:
**[Test build #106769 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/106769/testReport)** for PR 24795 at commit [`7f1753d`](https://github.com/apache/spark/commit/7f1753db2d15186399e5ef9d7ef88b2bc7eedf67).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/106769/
Test PASSed.

tgravescs reviewed (approved):
+1, I'll wait a bit longer to commit in case anyone else wants to review.

cloud-fan commented on sql/core/src/main/java/org/apache/spark/sql/execution/vectorized/WritableColumnVector.java:
```diff
@@ -604,7 +604,10 @@ public final int appendArray(int length) {
   public final int appendStruct(boolean isNull) {
     if (isNull) {
-      appendNull();
+      // This is the same as appendNull but without the assertion for struct types
```
why is this necessary?

revans2 replied:
Because `appendNull` itself has an assertion that you are not appending a struct.  So any call to `appendStruct` with `isNull` true would have failed.

cloud-fan commented on sql/core/src/test/scala/org/apache/spark/sql/SparkSessionExtensionSuite.scala:
```diff
@@ -251,6 +286,371 @@ object MyExtensions {
+   * Returns the result of evaluating this expression on the entire
+   * [[org.apache.spark.sql.vectorized.ColumnarBatch]]. The result of
+   * calling this may be a single [[org.apache.spark.sql.vectorized.ColumnVector]] or a scalar
+   * value. Scalar values typically happen if they are a part of the expression i.e. col("a") + 100.
```
I understand that this is in test not an API, but other people may look at this test to learn how to implement columnar operator, and I feel the current example is not that good.

IIUC, the goal is:
1. users can write a rule to replace an arbitrary SQL operator with a custom optimized columnar version
2. Spark automatically insert column-to-row and row-to-column operators around the columnar operator.

For 1, I think a pretty simple approach is, take in an expression tree, compile it to a columnar processor that can execute the expression tree in a columnar fashion. We don't need to create a `ColumnarExpression`.

revans2 replied:
The main reason it is this way is that originally I had Expression support columnar as well but as a part of the review it changed so I made minimal changes to the example to match.
I do find it simpler from a development standpoint to have a one to one mapping.  I can then write unit tests with just inputs and verify that the outputs match exactly.   But yes there probably are simpler ways to do this depending on the columnar library you are using.

cloud-fan commented:
Generally looks good, but I'm wondering if it's better to use Apache Arrow as the exchange format instead of the internal `ColumnarBatch`. I understand that this is a developer API so stability is not a concern here. I'm just afraid people may find `ColumnarBatch` hard to use when doing high-perf columnar operations, but they can't change it as it's in Spark.

tgravescs commented:
If you look at the discuss and vote threads there was discussion on this and people were worried about it since it wasn't 1.0 and could potentially change, so we wanted to separate that out and not do at this point

HyukjinKwon commented on sql/core/src/test/scala/org/apache/spark/sql/SparkSessionExtensionSuite.scala:
```diff
@@ -251,6 +286,371 @@ object MyExtensions {
     (_: Seq[Expression]) => Literal(5, IntegerType))
 }
 
+case class CloseableColumnBatchIterator(itr: Iterator[ColumnarBatch],
```
Not a big deal but for my own understanding, why is it `case class`?

revans2 replied:
It doesn't have to be a case class, just made it so I didn't need to add in the new, but I am happy to change it if you want me to in a follow on PR.

HyukjinKwon commented on sql/core/src/main/scala/org/apache/spark/sql/execution/WholeStageCodegenExec.scala:
```diff
@@ -840,34 +874,55 @@ case class CollapseCodegenStages(
+  private def adjustColumnar(plan: SparkPlan, isColumnar: Boolean): Boolean =
+    // We are walking up the plan, so columnar starts when we transition to rows
+    // and ends when we transition to columns
+  plan match {
```
not a biggie but indentation here looks weird.

HyukjinKwon commented:
Looks good to me in general too.
